### PR TITLE
Developing CLI for StandardBounties funding

### PR DESCRIPTION
**Pull request by lightclient**
@owocki please let me know how I am doing on issue #1. Both the interactive mode and json mode are working for me on `rinkeby` with ETH. I haven't tested EIP20 tokens with it yet, but the logic is there so if it isn't working it should only need a few minor tweaks.

### CODE REVIEW

**Comment by owocki** [comment]
looking at this now

File: src/input.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    # ensure the two required arguments are there
+    if( len(args.grouped.get('_')) != 2 ):
+        print('invalid number of arguments')
+        exit(1)
+
+    data.update({'issueURL' : args.grouped.get('_')[0]})
+    data.update({'amount'  : args.grouped.get('_')[1]})
+
+    # validate that issueURL is a url
+
```
Comment (owocki) [inline src/input.py]: TODO

File: src/utils/wallet.py [hunk 2]
```diff
@@ -0,0 +1,1473 @@
+"""This submodule provides the PublicKey, PrivateKey, and Signature classes.
+It also provides HDPublicKey and HDPrivateKey classes for working with HD
```
Comment (owocki) [inline src/utils/wallet.py]: is this an open source bit of code?  if so, maybe its worth using `pip` or at least putting attribution in it?

**Comment by owocki** [comment]
did not test, but code looks sane so far!

i was gonna use private keys not mnemonics, but i guess either works

IMHO we need to 
1. test with ERC20 tokens ( https://tokenfactory.surge.sh/ is great for this )
2. resolve all the todos
3. test again

File: src/utils/wallet.py (hunk 2 above)
Comment (lightclient) [reply src/utils/wallet.py]: Good catch, yes it open source. I forgot to add the attribution. I will see if it is available via pip or if something similar is.

**Comment by lightclient** [comment]
@owocki how would you prefer to interact with this tool? 

EDIT: I removed the json interface.

**Comment by lightclient** [comment]
CLI can fund bounties with EIP20 tokens and all TODOs are resolved! 😃

**Comment by lightclient** [comment]
Hi @owocki, I think this PR is ready to go. Please let me know if there is anything else to add / fix before I submit this work. Thank you!

**Comment by owocki** [comment]
testing now

**Comment by owocki** [comment]
1. defaults.json link in the readme is broken

2. so is this

> python3 funder/main.py https://github.com/gitcoinco/command_line_standardbounties_funder/issues/1 0.01

3. the schema for the input blob has actually cahnged in the last week or so.  could we add in these params?  https://github.com/gitcoinco/web/blob/master/app/assets/v2/js/pages/new_bounty.js#L204-L206

I get a stack trace trying to fund this 

```
(foo) kevinowocki@local /Users/kevinowocki/Sites/gitcoin/command_line_standardbounties_funder~ % python3 funder/cli.py https://github.com/gitcoinco/command_line_standardbounties_funder/issues/1 0.01

Github [owocki]:
Title:
Title: foo bar
Description: desc
Keywords: kd
Experience: sfdds
Error: invalid choice: sfdds. (choose from beginner, intermediate, advanced)
Experience: sfdsdf
Error: invalid choice: sfdsdf. (choose from beginner, intermediate, advanced)
Experience: beginner
Length: foo
Error: invalid choice: foo. (choose from hours, days, weeks, months)
Length: hours
Type:
Type: f
Error: invalid choice: f. (choose from bug, feature, security, other)
Type: bug

Saving data to IPFS... Traceback (most recent call last):
  File "funder/cli.py", line 89, in <module>
    main()
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "funder/cli.py", line 85, in main
    handler(state)
  File "/Users/kevinowocki/Sites/gitcoin/command_line_standardbounties_funder/funder/commands/issue_and_activate.py", line 85, in handler
    ipfsHash = saveToIPFS(state)
  File "/Users/kevinowocki/Sites/gitcoin/command_line_standardbounties_funder/funder/utils/ipfs.py", line 5, in saveToIPFS
    payload = buildPayload(data)
  File "/Users/kevinowocki/Sites/gitcoin/command_line_standardbounties_funder/funder/utils/ipfs.py", line 25, in buildPayload
    'experienceLevel': data.get('experience').capitalize(),
AttributeError: 'NoneType' object has no attribute 'capitalize'
```

4. could the system print out the ETH address its going to use at the top?  this would help me to fund the account being used with some ETH!

5. the final example in the readme creates a validation error

```
(foo) kevinowocki@local /Users/kevinowocki/Sites/gitcoin/command_line_standardbounties_funder~ % python3 funder/cli.py a 10.1 -ahf --github c-o-l-o-r --title title --description description --keywords key,word --token-address 0x4354321ef77766e2ec327ce58d3dff8358d46208

Usage: cli.py [OPTIONS] URL AMOUNT

Error: Invalid value for "url": url must be GitHub link
```

6. edit: yay the 2nd to last command worked!

```
(foo) kevinowocki@local /Users/kevinowocki/Sites/gitcoin/command_line_standardbounties_funder~ % python3 funder/cli.py https://github.com/gitcoinco/command_line_standardbounties_funder/issues/1 0.01 -bhf --github c-o-l-o-r

Title: foo the bar
Description: foo the bar
Keywords: lol

Saving data to IPFS... QmYuxqBY3pij3bnhGCC2tGtEbMNsS4cxNHbLU3LER83wX9
Funding bounty... 0x382f8ed2a3740d23f330b965ef62e25239e797568083aabd9708b5eaa99dcd45
Bounty 832 funded successfully!
```

7. Tokens can be specified using --token GIT where GIT

just to avoid any confusion (Gitcoin is not a token), could we set a different example token?

**Comment by owocki** [comment]
8. `keywords` should be a string of keywords that are comma delimited, not an array

**Comment by owocki** [comment]
9. have you tested with an ERC20 token? i can't seem to get that functionality working.  it keeps trying to send ETH

**Comment by owocki** [comment]
signing off for the night.  sorry for the dribble of comments.  talk to you lata!

**Comment by lightclient** [comment]
Thanks for taking a look at it! I made several of the changes you mentioned. I'm currently investigating number 3. It seems like if you don't enter an expected value for the first prompt it sets the value to `None`. I'll be back online tomorrow night and get it fixed 😄

**Comment by owocki** [comment]
thanks! ping me when i should test again

**Comment by lightclient** [comment]
@owocki that bug should be fixed. It's read to be tested again whenever you get a chance. I tested ETH and ERC20 bounties on Rinkeby and they were both successful for me.

**Comment by owocki** [comment]
looking now

**Comment by owocki** [comment]
eth bounties working..

i get this error when i try to fund with an ERC20 token

```
(foo) kevinowocki@local /Users/kevinowocki/Sites/gitcoin/command_line_standardbounties_funder~ % python3 funder/cli.py https://github.com/gitcoinco/command_line_standardbounties_funder/issues/1 0.1 -ahf --github c-o-l-o-r --title title --description description --keywords key,word --confirm --token-address 0x6549087d9d06d6275fb59900208c22dc4ab2e6d8

Funding from 0x627306090abab3a6e1400e9345bc60c78a8bef57
Getting token info... Traceback (most recent call last):
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/web3/contract.py", line 1363, in call_contract_function
    output_data = decode_abi(output_types, return_data)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/eth_abi/abi.py", line 96, in decode_abi
    return decoder(stream)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/eth_abi/decoding.py", line 118, in __call__
    return self.decode(stream)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/eth_utils/functional.py", line 22, in inner
    return callback(fn(*args, **kwargs))
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/eth_abi/decoding.py", line 164, in decode
    yield decoder(stream)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/eth_abi/decoding.py", line 118, in __call__
    return self.decode(stream)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/eth_abi/decoding.py", line 133, in decode
    start_pos = decode_uint_256(stream)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/eth_abi/decoding.py", line 118, in __call__
    return self.decode(stream)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/eth_abi/decoding.py", line 186, in decode
    raw_data = self.read_data_from_stream(stream)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/eth_abi/decoding.py", line 279, in read_data_from_stream
    len(data),
eth_abi.exceptions.InsufficientDataBytes: Tried to read 32 bytes.  Only got 0 bytes

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "funder/cli.py", line 100, in <module>
    main()
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "funder/cli.py", line 96, in main
    handler(state)
  File "/Users/kevinowocki/Sites/gitcoin/command_line_standardbounties_funder/funder/commands/issue_and_activate.py", line 92, in handler
    state.update(getTokenInfo(state))
  File "/Users/kevinowocki/Sites/gitcoin/command_line_standardbounties_funder/funder/utils/token.py", line 28, in getTokenInfo
    'name': token.functions.symbol().call(),
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/web3/contract.py", line 1106, in call
    **self.kwargs
  File "/Users/kevinowocki/code/.virtualenvs/foo/lib/python3.6/site-packages/web3/contract.py", line 1385, in call_contract_function
    raise BadFunctionCallOutput(msg) from e
web3.exceptions.BadFunctionCallOutput: Could not decode contract function call symbol return data b'' for output_types ['string']
```

**Comment by owocki** [comment]
mind dropping in the command you used to test?

**Comment by owocki** [comment]
let me know about number 3 also

**Comment by lightclient** [comment]
Hey @owocki, here is the command I'm using to fund with ERC20 tokens:

```
python3 funder/cli.py https://github.com/gitcoinco/command_line_standardbounties_funder/issues/1 0.1 -ahf --github c-o-l-o-r --title title --description description --keywords key,word --token-address 0x4354321ef77766e2ec327ce58d3dff8358d46208
```

The script is currently defaulting to use Rinkeby. I don't see your token deployed on it? https://rinkeby.etherscan.io/address/0x6549087d9d06d6275fb59900208c22dc4ab2e6d8

Please try the command from above using same Ethereum address. I sent 1 BUG token to it!

As for the bug in 3 it should be fixed now 😄

**Comment by owocki** [comment]
yayyyy it works 

```
(foo) kevinowocki@local /Users/kevinowocki/Sites/gitcoin/command_line_standardbounties_funder~ % python3 funder/cli.py https://github.com/gitcoinco/command_line_standardbounties_funder/issues/1 0.1 -ahf --github c-o-l-o-r --title title --description description --keywords key,word --token-address 0x4354321ef77766e2ec327ce58d3dff8358d46208


Funding from 0x627306090abab3a6e1400e9345bc60c78a8bef57
Getting token info... done.
Token balance: 1.0
Ether balance: 0.5668782183000001
Do you want to continue? [y/N]: y
Saving data to IPFS... QmSAqcuwXU5aGKq4q9f9BYMwkjYx4sNmGQgqmRFFztYQGu
Approving token usage...

0x0a3b09d475880705f2ce056e84b3fe496b2c5193d046e280c469a84483b87f9f
Funding bounty... 0x69541400cbb9e42c6b978aea99cdb09716fe8c14444d827337416b642492dcef
Bounty 854 funded successfully!

```

**Pull request merged**